udsoncan/services/Authentication.py

```diff
@@ -0,0 +1,383 @@
+        if authentication_task in (Authentication.AuthenticationTask.deAuthenticate,
+                                   Authentication.AuthenticationTask.authenticationConfiguration):
+            # Service ID + Subfunction, no param
+            pass
+        elif authentication_task in (Authentication.AuthenticationTask.verifyCertificateUnidirectional,
+                                     Authentication.AuthenticationTask.verifyCertificateBidirectional,
+                                     Authentication.AuthenticationTask.requestChallengeForAuthentication):
+            tools.validate_int(communication_configuration, min=0, max=0xFF, name='Communication Configuration')
+            # communicationConfiguration
+            data = struct.pack('B', communication_configuration)
+            if authentication_task in (Authentication.AuthenticationTask.verifyCertificateUnidirectional,
+                                       Authentication.AuthenticationTask.verifyCertificateBidirectional):
+                # certificateClient
+                data = Authentication._append_byes_parameter(data, certificate_client, 'Certificate Client')
+
+                # challengeClient
+                data = Authentication._append_byes_parameter(data, challenge_client, 'Challenge Client')
+            else:
+                if not isinstance(algorithm_indicator, bytes) or len(algorithm_indicator) != 16:
+                    raise ValueError(f'{algorithm_indicator} must be a bytes object of length 16')
+                # algorithmIndicator
+                data += algorithm_indicator
+        elif authentication_task == Authentication.AuthenticationTask.proofOfOwnership:
+            data = b''
+            # proofOfOwnershipClient
+            data = Authentication._append_byes_parameter(data, proof_of_ownership_client, 'Proof Of Ownership Client')
+            # ephemeralPublicKeyClient
+            data = Authentication._append_byes_parameter(data, ephemeral_public_key_client, 'Ephemeral Public Key'
+                                                                                            ' Client')
+
+        elif authentication_task == Authentication.AuthenticationTask.transmitCertificate:
+            tools.validate_int(certificate_evaluation_id, min=0, max=0xFFFF, name='Certificate Evaluation ID')
+            # certificateEvaluationId
+            data = struct.pack('>H', certificate_evaluation_id)
+            # certificateData
+            data = Authentication._append_byes_parameter(data, certificate_data, 'Certificate Data')
+
+        elif authentication_task in (Authentication.AuthenticationTask.verifyProofOfOwnershipUnidirectional,
+                                     Authentication.AuthenticationTask.verifyProofOfOwnershipBidirectional):
+            if not isinstance(algorithm_indicator, bytes) or len(algorithm_indicator) != 16:
+                raise ValueError(f'{algorithm_indicator} must be a bytes object of length 16')
```

Comment:
Same here:

`udsoncan/services/Authentication.py:246: error: If x = b'abc' then f"{x}" or "{}".format(x) produces "b'abc'", not "abc". If this is desired behavior, use f"{x!r}" or "{!r}".format(x). Otherwise, decode the bytes`

Reply:
will fix